### DISCUSSION TITLE
Import as "fully rendered HTML"

### DISCUSSION THREAD

**Discussion by ctft1** [TriliumNext/Trilium#4260] [Ideas]
Hi,

When importing a HTML file, it's either:
 - converted as simplified markup, which is usually fine.
 - or not rendered if imported without the HTML processing (can be then "Opened with" the browser)

But sometimes it can be useful to have a more faithful copy of the HTML, especially on more complex websites (ajax, ..) that needs to be kept for archive.

I've been looking at singlefile2trilium script, which can work but needs python running.

Would it be possible to have a dedicated "Import as fully rendered HTML" option? This would most likely :
- Save the HTML file as a note (without modifying it)
- Create the rendered note with the iframe tag, and the right reference to the HTML note

Thank you!

regards,

**Comment by sottey** [comment]
The problem with iFrames is that, "html...that needs to be kept for archive" is that it isn't actually capturing anything. It is just creating a "browser within Trilium". So when the site goes away, the iframe doesn't load.

I have done about a month's worth of work just focusing on how to integrate an existing codebase into Trilium. While it certainly takes some work, it is ultimately more effective than relying on an outside party.

Of course, this only works with Standalone Javascript apps. Things that are much more complex are not possible (or are too work intensive to be worthwhile).

Is there a specific site you were hoping to include with Trilium?

**Reply by Nriver** [reply]
Singlefile gives you a complete "clone" of what you see in the browser into a single html file. It works offline. The iframe is just a way to show the content in Trilium.

This is what it looks like in Trilium. Pixel perfect copy :)
![ksnip_20230914-094746](https://github.com/zadam/trilium/assets/6752679/68057cb3-5988-480a-b158-255e4f213ab2)

**Reply by ctft1** [reply]
Are you using the singlefile2trilium scripts or another method?
I would like to find a quick way to get this result, but without the python script running to check for saved files (I'm using Trilium as a portable app, so I don't always have python + my folder path can change)

I was also thinking about asking the SingleFile team to add a new "save destination" option that could be to customize a REST api call. That would be (kind of) easy on their side, and the singlefile2trilium JS (and listener) would stay on the Trilium side..

**Reply by Nriver** [reply]
The same way as you do. It would be nice if SingleFile can directly communicate with Trilium's api.

**Comment by ctft1** [comment]
In my use case I'm importing HTML files that have been created with Singlefile extension. I like the way Singlefile includes everything in one HTML file, it's pretty easy to maintain and archive. Plus it's really efficient in capturing pages as you see them.

That means the webpage content would be fixed (as it was scrapped), the iframe idea was more about having a way to fully render the page (otherwise when importing, its content is simplified with MD)

Right now I'm using Trilium + some Firefox addons (Scrapbee, ex-Scrapbook Plus, and also SingleFile) for storing some webpages, but it's not centralized...

My perfect usage for Trilium would be to handle my global personal "knowledge base", mixing and organizing notes, todos, code, but also some contents like html pages, pdf, ...  It means something mixing Notes and Document Management.

Example : If I plan to buy a house, it's nice not being able to only take notes, but also save static contents for some listings (HTML pages fully rendered with JS as it's everywhere), raw pictures, putting all of this in a structured way like Trilium deals perfectly with..

The direct import already works for PDF or images (a new note is created and content can be seen inside Trillium), but it's not possible with HTML files without loosing some content when simplified (a workaround is to manually import with correct options, then link the note with another one having an iframe to render it correctly) .. Maybe I missed something trivial in my use case that could do that?

Of course I can also do it with a screenshot, but it's not the same for searching the page + also not easy to handle with long content pages

**Comment by zerebos** [comment]
After reading through this thread, I'm working on an addon for Trilium that should improve this experience a bit and not require python running or anything external aside from using the SingleFile extension itself.

This would satisfy these requirements that you laid out:
- Native(ish) to Trilium
- No external (python) running program
- SingleFile html saved as a single separate note
- `<iframe>` referencing the SingleFile html note
- One-time import option and interface

And similar to `singlefile2trilium` it would also have:
- Auto import newly downloaded SingleFile pages (configurable)
- Original information saved as attributes

And some features specific to this addon:
- Full settings panel inside Trilium
- Folder watching via Trilium Desktop
- One-time import button in the UI
- (potentially) Enable searching inside `<iframe>` properly


Here's a preview of how this page looks imported:
![image](https://github.com/zadam/trilium/assets/6865942/5999af19-a67d-43a5-b943-717c28c687c5)

**Reply by ctft1** [reply]
That would be amazing! Any idea for a first preview date?

**Reply by zerebos** [reply]
Probably not today as I'm pretty distracted by football, but definitely sometime this week! For now, I'll give you another sneak peek of how the addon looks so far:

https://github.com/zadam/trilium/assets/6865942/8f6dd790-18ac-49bb-b494-247cd8edfa30

**Reply by zerebos** [reply]
It ended up being less work than I thought to fix the things I needed to fix. If you'd like to give it a try I have installation instructions (and the download itself) over here: https://github.com/rauenzi/Trilium-SingleFile

I'd love to get some feedback, as the way I've done things is a bit opinionated, so I would need user feedback to adjust it and see what should be able to be customized. And of course, I'm sure there will be plenty of bugs.

**Reply by Nriver** [reply]
It's magic!

**Reply by zerebos** [reply]
Hopefully the good kind! 😄 I also just updated the README to add some info about `#singleFileInbox` and where imports are placed since I don't think that's clear from using the addon itself.

**Reply by ctft1** [reply]
That looks amazing, thank you!

When importing the ZIP itself, should we only uncheck the "Safe Import" or should we also uncheck some other boxes ?
I still need some testing, right now I cannot find my imported files (not sure I fully understand the #singleFileInbox or #inbox)

Also the import / setup page is not well rendered with light theme (text is not seeable as in white I guess)

**Reply by zerebos** [reply]
"Safe Import" is the only one you need to uncheck for it to import properly yes. When you import a new SingleFile page it will try to import it under a note with the `#singleFileInbox` attribute first. If that doesn't exist, it tries to use a note with `#inbox`. If that doesn't exist then it uses your daily note.

**Reply by ctft1** [reply]
Great, actually that's what I was doing. And now I realize the issue was more with an empty JSON setting when installing. Trying to drag and drop a HTML immediately after importing the ZIP seems to work (red then green button), but nothing happens (no note is created). 

Workaround is : go in Settings, and change at least 1 setting. At this time the JSON is populated, and the drag and drop works perfectly.

And last : I thought my import had failed because of the Light Theme issue: everything on the 2 tabs is dark text on dark background... it means I could not really see the tabs and settings.

This how it looks like in Light Theme: 

![image](https://github.com/zadam/trilium/assets/95131518/a0988dfd-4034-4aa6-9bd7-af209f768a7b)

**Reply by zerebos** [reply]
Huh weird, you must have come high specificity in your theme, I will work on this.

**Comment by jwhonce** [comment]
@rauenzi I installed the plugins for Chrome, Brave and Safari saved files, imported saved SingeFile into linux and macosx clients everything worked perfectly. Thanks for this. 

PS I followed your example for SingleFile settings. Your screenshot is appears to be behind the current extension.

PPS Sometimes drag and drop to linux client does not work. The dashed box does not go bold  and dropping does nothing. Reloading the front end fixes the issue and I was back in business.

**Reply by zerebos** [reply]
Yeah no problem, I'm glad it's working out!

> BTW I followed your example for SingleFile settings. Your screenshot is appears to be behind the current extension.

Not sure exactly what you mean by this though. I didn't provide any screenshots of SingleFile extension settings. Maybe you're talking about `singlefile2trilium.py` instead?

**Reply by jwhonce** [reply]
Now I am embarrassed that is what I get for trying three solutions on the same afternoon. :-) You are correct.

**Reply by jwhonce** [reply]
I should have added the linux client is running on fedora out of flatpak

**Reply by zerebos** [reply]
> PPS Sometimes drag and drop to linux client does not work. The dashed box does not go bold and dropping does nothing. Reloading the front end fixes the issue and I was back in business.

Have you noticed any pattern with this? or any errors in console?

**Reply by jwhonce** [reply]
Haven't yet, but will update if I do.
